### PR TITLE
Update starcgen for new types

### PR DESCRIPTION
**Pull request by lostluck**
Updates the starcgen code generator to better handle Pointer, Map, and Array types.

* Slice suffix for function names is now a prefix to avoid collisions in ordering with Maps of Slice of Arrays or similar
* Filter out "universal" named types such as error.
* Filter out Beam internal types (beam.X/typex.X and similar)
* Improve test coverage for both the templates and the analyzer.
* Regenerate the beam shims with `go generate ./...`

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by lostluck** [comment]
R: @aaltay Please review. 
The material changes are in the files under sdks/go/pkg/beam/util/. The rest are just the consequent generated code changes.

File: sdks/go/pkg/beam/util/shimx/generate.go [hunk 1]
```diff
@@ -176,14 +176,27 @@ type Wrap struct {
 // See https://golang.org/ref/spec#Identifiers for details.
 func Name(t string) string {
 	if strings.HasPrefix(t, "[]") {
-		return Name(t[2:] + "Slice")
+		return "SliceOf" + Name(t[2:])
+	}
+	if strings.HasPrefix(t, "map[") {
+		if i := strings.Index(t, "]"); i >= 0 {
+			// It should read MapOfTypeName.
```
Comment (aaltay) [inline sdks/go/pkg/beam/util/shimx/generate.go]: MapOfTypeName -> MapOfKeyTypeName_ValueTypeName ?

**Comment by lostluck** [comment]
Run Go PreCommit

File: sdks/go/pkg/beam/util/shimx/generate.go (hunk 1 above)
Comment (lostluck) [reply sdks/go/pkg/beam/util/shimx/generate.go]: Done. Nice catch!

**Comment by lostluck** [comment]
The Jenkin's environment appears to be interestingly configured due to GoGradle, so the tests for the shim-generator are unable to handle imports reliably there. They work fine locally.

I've opted to add special casing to check if that code can succeed first, and skip the test if not. It's sufficient that they run every time I do `go test ./...` locally.

**Pull request merged**